src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,354 @@
+
+            var headersInCache = new Dictionary<int, ProvenBlockHeader>();
+
+            var provenHeadersOutput = new SortedDictionary<int, ProvenBlockHeader>();
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                int index = fromBlockHeight;
+
+                ProvenBlockHeader header = null;
+
+                var blockHeightsNotInCache = new List<int>();
+
+                do
+                {
+                    if (this.Cache.TryGetValue(index, out header))
+                    {
+                        headersInCache.Add(index, header);
+                    }
+                    else
+                    {
+                        blockHeightsNotInCache.Add(index);
+                    }
+
+                    index++;
+
+                } while (index <= toBlockHeight);
+
+                // Try and get items from the repository if not found in the store cache.
+                if (blockHeightsNotInCache.Count > 0)
+                {
+                    var repositoryHeaders = new Dictionary<int, ProvenBlockHeader>();
+
+                    // If headersInCache is empty then we can assume blockHeightsNotInCache is the full range.
+                    if (headersInCache.Keys.Count == 0)
+                    {
+                        List<ProvenBlockHeader> rangeHeaders = await this.provenBlockHeaderRepository.GetAsync(fromBlockHeight, toBlockHeight);
+
+                        index = fromBlockHeight;
+
+                        foreach (ProvenBlockHeader rangeHeader in rangeHeaders)
+                        {
+                            if (rangeHeader != null)
+                            {
+                                repositoryHeaders.Add(index, rangeHeader);
+                                this.Cache.AddOrUpdate(index, rangeHeader, rangeHeader.HeaderSize);
+                            }
+
+                            index++;
+                        }
+                    }
+                    else
+                    {
+                        // If not a full sequence then check individually.
+                        foreach (int headerNotInCache in blockHeightsNotInCache)
+                        {
+                            ProvenBlockHeader repositoryHeader = await this.provenBlockHeaderRepository.GetAsync(headerNotInCache).ConfigureAwait(false);
+
+                            if (repositoryHeader != null)
+                            {
+                                repositoryHeaders.Add(headerNotInCache, repositoryHeader);
+
+                                this.Cache.AddOrUpdate(headerNotInCache, repositoryHeader, repositoryHeader.HeaderSize);
+                            }
+                        }
+                    }
+
+                    // Before returning the items, make sure they are sorted.
+                    provenHeadersOutput = AddToSortedDictionary(headersInCache, provenHeadersOutput);
```

Comment:
This does not guarantee sequantialness it only sorts the items, what if you have items 5, 7, 8 ?

Reply:
Ok, done.